src/execution/operator/order/physical_top_n.cpp

```diff
@@ -251,13 +132,93 @@ TopNHeap::TopNHeap(ExecutionContext &context, const vector<LogicalType> &payload
     : TopNHeap(context.client, Allocator::Get(context.client), payload_types, orders, limit, offset) {
 }
 
-void TopNHeap::Sink(DataChunk &input) {
-	sort_state.Sink(input);
+void TopNHeap::Sink(DataChunk &input, optional_ptr<TopNBoundaryValue> global_boundary) {
+	// compute the ordering values for the new chunk
+	sort_chunk.Reset();
+	executor.Execute(input, sort_chunk);
+
+	// construct the sort key from the sort chunk
+	vector<OrderModifiers> modifiers;
+	for (auto &order : orders) {
+		modifiers.emplace_back(order.type, order.null_order);
+	}
+	sort_keys.Reset();
+	auto &sort_keys_vec = sort_keys.data[0];
+	CreateSortKeyHelpers::CreateSortKey(sort_chunk, modifiers, sort_keys_vec);
+
+	// fetch the current global boundary (if any)
+	string boundary_val;
+	string_t global_boundary_val;
+	if (global_boundary) {
+		boundary_val = global_boundary->GetBoundaryValue();
+		global_boundary_val = string_t(boundary_val);
+	}
+
+	// insert the sort keys into the priority queue
+	constexpr idx_t BASE_INDEX = NumericLimits<uint32_t>::Maximum();
+
+	bool any_added = false;
+	auto sort_key_values = FlatVector::GetData<string_t>(sort_keys_vec);
+	for (idx_t r = 0; r < input.size(); r++) {
+		auto &sort_key = sort_key_values[r];
+		if (!boundary_val.empty() && sort_key > global_boundary_val) {
```

Comment:
> We could remedy that by adding templated heaps for primitive types in the future.

Potentially, we could get a lot of the benefit without templating by recognizing when the sort keys have a fixed size and using a fixed size `memcmp` instead of the regular string comparison. The less-than comparison could look like this: `FastMemcmp(lhs.GetPrefix(), rhs.GetPrefix(), fixed_size) < 0` (if the fixed size is less than the inline length). This should give a pretty good performance improvement for comparisons without having to use a lot of templating.

Reply:
Thanks for the suggestion! I gave it a shot [here](https://github.com/Mytherin/duckdb/commit/ca4f0f95ecd4c86d4d0e4aadd6a7c228a4b7809b), but the difference in end-to-end performance is negligible at least on the above queries. Profiling this it seems the main performance cost is in constructing the sort keys - and not the comparisons themselves.

I think to get optimal performance for simple keys we'll need to template this, but I'll leave that for a future PR. I think the difference in performance is not large enough to worry about.